Convert to source generator

ycanardeau opened the pull request:
Turns the code generation into a source generator.

References
- [【Roslynメタプログラミング】ValueChangedGanerator | ++C++; // 未確認飛行 C ブログ](https://ufcpp.net/blog/2015/6/valuechangedganerator/)
- [New C# Source Generator Samples | .NET Blog](https://devblogs.microsoft.com/dotnet/new-c-source-generator-samples/)
- [YairHalberstadt/stronginject: compile time dependency injection for .NET](https://github.com/YairHalberstadt/stronginject)

By the way, I enjoy watching your YouTube videos and find them to be very helpful. :) I used your videos as a reference to write [VocaDB.ResXFileCodeGenerator](https://github.com/VocaDB/ResXFileCodeGenerator). Thank you!

ufcpp commented:
memo (may be my tasks):

- [ ] add package info like [this](https://github.com/ufcpp/StringLiteralGenerator/blob/master/src/StringLiteralGenerator/StringLiteralGenerator.csproj#L9-L15)
- [ ] add CI workflow like [this](https://github.com/ufcpp/StringLiteralGenerator/blob/master/.github/workflows/build.yml)
- [ ] SourceGenerator should append text directly into `StringBuilder` instead of `SyntaxNode.ToFullString` for performance reason
- [x] add tests for SourceGenerator
- [ ] use [Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.XUnit](https://github.com/dotnet/roslyn-sdk/blob/master/src/Microsoft.CodeAnalysis.Testing/README.md) to test ValueChangedGeneratorCodeFixProvider
- [ ] update documents:
    - describe about SourceGenerator
    - how different Code Fix and SourceGenerator are

ufcpp commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.SourceGenerator/RoslynExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+﻿using System.Collections.Generic;
```
How about using the following `Compile` tags instead of copy-and-pasting files

```xml
  <ItemGroup>
    <Compile Include="../ValueChangedGenerator.CodeFixes/RecordDefinition.cs" />
    <Compile Include="../ValueChangedGenerator.CodeFixes/SyntaxExtensions.cs" />
  </ItemGroup>
```

ycanardeau commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.Test/ValueChangedGeneratorUnitTests.cs:
```diff
@@ -19,7 +22,7 @@ public async Task TestMethod1()
         //Diagnostic and CodeFix both triggered and checked for
-        [TestMethod]
+        [Fact]
         public async Task TestMethod2()
```
For your information, TestMethod2 was already broken when I forked the repo. Could you please check it when you have some time?

ufcpp replied:
Not for a while...

ufcpp replied:
❌ Not for a while
⭕ Not at all

`TestMethod1` and `TestMethod2` are default codes in the "Analyzer with Code Fix" template. These are useless for the ValueChangedGenerator.

ufcpp replied:
As far as using CSharpCodeFixVerifier, it is difficult to test code fix which generates additional C# files. This use case is out of CSharpCodeFixVerifier's scope. Maybe we should give up testing the Code Fix (and recommend users to migrate to the Source Generator).

ycanardeau commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.SourceGenerator/Generator.cs:
```diff
@@ -0,0 +1,132 @@
+            var def = new RecordDefinition(strDecl);
+            var generatedNodes = GetGeneratedNodes(def).ToArray();
+
+            var newClassDecl = container.GetContainingTypesAndThis()
+                .Select((type, i) => i == 0
+                    ? ClassDeclaration(type.Name).GetPartialTypeDelaration().AddMembers(generatedNodes)
+                    : ClassDeclaration(type.Name).GetPartialTypeDelaration())
+                .Aggregate((a, b) => b.AddMembers(a));
```
I think there should be a better way to support nested partial classes, but I'll keep this as is for now.

ycanardeau commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.Test/ValueChangedGenerator.Test.csproj:
```diff
@@ -1,27 +1,33 @@
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.Analyzer.Testing.MSTest" Version="1.0.1-beta1.*" />
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.CodeFix.Testing.MSTest" Version="1.0.1-beta1.*" />
-    <PackageReference Include="Microsoft.CodeAnalysis.VisualBasic.CodeRefactoring.Testing.MSTest" Version="1.0.1-beta1.*" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="3.8.0" />
+    <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.2" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.Analyzer.Testing.XUnit" Version="1.0.1-beta1.21109.2" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.CodeFix.Testing.XUnit" Version="1.0.1-beta1.21109.2" />
+    <PackageReference Include="Microsoft.CodeAnalysis.CSharp.CodeRefactoring.Testing.XUnit" Version="1.0.1-beta1.21109.2" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.7.1" />
+    <PackageReference Include="xunit" Version="2.4.1" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.3">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
+    <PackageReference Include="coverlet.collector" Version="1.3.0">
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+      <PrivateAssets>all</PrivateAssets>
+    </PackageReference>
```
Replaced MSTest with xUnit.

ycanardeau commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.Test/ValueChangedGeneratorUnitTests.cs:
```diff
@@ -55,5 +58,214 @@ class TYPENAME
+            public int X;
+            public int Y;
+            public int Z => X * Y;
+
+            /// <summary>
+            /// Name.
+            /// </summary>
+            public string Name;
+        }
+
+        public partial class C3 : BindableBase
+        {
+            struct NotifyRecord
+            {
+                public int X;
+                public int Y;
+                public int Z => X * Y;
+
+                /// <summary>
+                /// Name.
+                /// </summary>
+                public string Name;
+            }
+        }
+    }
+}";
+            var compilation = CSharpCompilation.Create(assemblyName: "test", options: new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary))
+                .AddSyntaxTrees(CSharpSyntaxTree.ParseText(text));
+            var generator = new Generator();
+            var syntaxTrees = compilation.SyntaxTrees;
+            Assert.Single(syntaxTrees);
+            var classDecl = (ClassDeclarationSyntax)syntaxTrees.First().GetCompilationUnitRoot()
+                .DescendantNodes().First(x => x is ClassDeclarationSyntax)
+                .DescendantNodes().First(x => x is ClassDeclarationSyntax)
+                .DescendantNodes().First(x => x is ClassDeclarationSyntax);
+            var model = compilation.GetSemanticModel(classDecl.SyntaxTree);
+            var container = model.GetDeclaredSymbol(classDecl);
+            Assert.NotNull(container);
+            Assert.Equal(expected, actual: generator.GeneratePartialDeclaration(container!, classDecl).ToFullString());
+        }
```
I'm really not sure if this is the right way to write tests, but I tried my best.

ufcpp replied:
I'm also not sure. Source Generators are not so mature. This code looks good so far.

ycanardeau reviewed (commented):
Thank you for reviewing my PR!

ycanardeau commented:
> SourceGenerator should append text directly into `StringBuilder` instead of `SyntaxNode.ToFullString` for performance reason

Do you mean that I should use `SourceText.From` just like you did in [StringLiteralGenerator](https://github.com/ufcpp/StringLiteralGenerator/blob/b02f55493254d42391adbb811c11b175b5088816/src/StringLiteralGenerator/Utf8StringLiteralGenerator.cs#L51)?

ufcpp commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.Test/Verifiers/CSharpAnalyzerVerifier`1+Test.cs:
```diff
@@ -0,0 +1,26 @@
+﻿using Microsoft.CodeAnalysis.CSharp.Testing;
```
These files in Verifiers folder can be removed. There exist equivalent codes in the Assembly Microsoft.CodeAnalysis.Analyzer.Testing and Assembly Microsoft.CodeAnalysis.CodeFix.Testing packages.

ufcpp commented:
> Do you mean that I should use SourceText.From just like you did in StringLiteralGenerator?

Yes but this has both pros and cons.

Advantages:

- performance

Drawbacks

- duplication
- difficult to format

ufcpp commented on ValueChangedGenerator/ValueChangedGenerator/ValueChangedGenerator.Test/ValueChangedGeneratorUnitTests.cs:
```diff
@@ -1,4 +1,8 @@
-﻿using System.Threading.Tasks;
+﻿using System.Linq;
```
Needs one more test case: class in namespace

ycanardeau commented:
Removed redundant tests and added a test for nested namespaces. Perhaps we could replace `SyntaxNode.ToFullString` with `StringBuilder` later. It seems that [CsWin32](https://github.com/microsoft/CsWin32/blob/14796cd776e31436bfce4d09519ffacf47363871/src/Microsoft.Windows.CsWin32/SourceGenerator.cs#L148) uses `ToFullString` by the way.
